proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,278 @@
+to use the dapp. They will be applied even if transaction fails contrary to lamport transfers.
+
+## Detailed Design
+
+As a high-performance cluster, we want to incentivize players which feed the network
+with correctly formed transactions instead of spamming the network. The introduction
+of application fees would be an interesting way to penalize the bad actors and dapps
+can rebate these fees to the good actors. This means the dapp developer has to decide who to
+rebate and who to penalize special instructions. In particular, it needs to be able to
+penalize, even if the application is not CPI'd to. There were multiple possible approaches
+discussed to provide the application with access to transfer lamports outside of the regular cpi
+execution context. The following approach seems the best.
+
+*Passing the application fees in the instruction for each account and validating in the dapp*. A `PayApplicationFee`
+instruction is like irrevocable transfer instruction and will do the transfer even if the transaction fails. A new
+instruction `CheckApplicationFee` will check if the application fee for an account has been paid.
+
+    1. High degree of flexibility for programs (**++**)
+    2. No need to save the application fees on the ledger. (**++**)
+    3. Each transaction has to include additional instructions so it will break existing dapp interfaces (**-**)
+    4. Account structure does not need to be modified (**++**)
+    5. Does not prevent any denial of service attack on a smart contract (**-**)
+    6. Each dapp has to implement a cpi to check if a transaction has paid app fees
+
+If existing dapps like openbook want to implement application fees then all the dapps depending on openbook 
+also have to do additional development. The checks on the application fees will be responsibility of dapp developers.
+The plus point is that we do not have to touch the core solana structure to store the application fees.
+It is also more easier to calculate total fees.
+
+### A new application fee program
+
+We add a new native solana program called application fees program with program id.
+
+```
+App1icationFees1111111111111111111111111111
+```
+
+This native program will be used to check application fee for an account, to intialize rebates
+initaited by the account authority, and a special instruction by which fee payer accepts
+amount of lamports they are willing to pay as application fees per account.
+
+#### PayApplicationFees Instruction
+
+With this instruction, the fee payer accepts to pay application fees specifying the amount.
+This instruction **MUST** be included in the transaction that interacts with dapps having application fees.
+This instruction is like an irrevokable transfer if the payer has enough funds
+i.e even if the transaction fails the payer will end up paying.
+If this instruction is added then even if the dapp does not check the application fees the payer ends
+up paying.
```

Comment:
Not sure if this should be a concern, just think out loud here. With `PayApplicationFee` ix, L1 collects fee from user, transfer to dapp accounts. But it is dapps to decide if good/bad actors, if to rebate, and at what rate. Feels like the responsibility for the fund withdrew from user are not super clear. Perhaps need some tooling to help user understand where the fund distributed and rebated etc.

Reply:
Ok i will make it more clear.